source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,389 @@
+ * even one time hash searching.
+ *
+ * This is useful when some keys are frequently used and the keys are known at compile time or
+ * bootstrap time. You can get superior performance by using the inline handle. For example, the
+ * filter state always uses the filter name as the key and the filter name is known at compile time.
+ * By using the inline handle, the filter state could get the key/value pair without any hash.
+ *
+ * This class also provides the interface to access the key/value pair by the normal key and the
+ * interface has similar searching performance as the normal hash map. But the insertion and removal
+ * by the normal key is slower than the normal hash map.
+ *
+ *
+ * These is usage example:
+ *
+ *   // Define registry to store the inline keys. The registry should have lifetime that no shorter
+ *   // than all inline maps created by the registry.
+ *   InlineMapRegistry registry;
+ *
+ *   // Register the inline key. We should never do this after bootstrapping. Typically, we can
+ *   // do this by define a global variable.
+ *   auto inline_handle = registry.registerInlineKey("inline_key");
+ *
+ *   // Finalize the registry. No further changes are allowed to the registry after this point.
+ *   registry.finalize();
+ *
+ *   // Create the inline map.
+ *   auto inline_map = registry.createInlineMap<std::string>();
+ *
+ *   // Get value by inline handle.
+ *   inline_map->insert(inline_handle, std::make_unique<std::string>("value"));
+ *   EXPECT_EQ(*inline_map->lookup(inline_handle), "value");
+ */
+class InlineMapRegistry {
+public:
+  /**
+   * This is the handle used to access the key/value pair in the inline map. The handle is
+   * lightweight and could be copied and passed by value.
+   */
+  class InlineKey {
+  public:
+    /**
+     * Get the id of the inline map registry that the handle created by. This will be used
+     * to determine if the handle could be accepted by a inline map.
+     */
+    uint64_t registryId() const { return registry_id_; }
+
+    /**
+     * Get the id of the inline entry in the inline map. This could be used to access the
+     * key/value pair in the inline map.
+     */
+    uint64_t inlineId() const { return inline_id_; }
+
+    /**
+     * Get the key view of the inline key.
+     */
+    absl::string_view inlineKey() const { return inline_key_; }
+
+  private:
+    friend class InlineMapRegistry;
+
+    // This constructor should only be called by InlineMapRegistry.
+    InlineKey(uint64_t registry_id, uint64_t inline_id, absl::string_view inline_key)
+        : registry_id_(registry_id), inline_id_(inline_id), inline_key_(inline_key) {}
+
+    uint64_t registry_id_{};
+    uint64_t inline_id_{};
+    absl::string_view inline_key_;
+  };
+
+  // This is the Hash used for registration map/set and underlay hash map.
+  using Hash = absl::container_internal::hash_default_hash<std::string>;
```

Comment:
why do you need this? It wasn't needed in my sketch of how this should work.

Reply:
This `Hash` is used to compute hash of key manually. Then we needn't do that repeatly when lookup  a normal key.

check the `lookupImpl`.

Reply:
I would not do that. Keep it simple!

Reply:
done.